modules/gdscript/language_server/gdscript_language_protocol.cpp

```diff
@@ -32,32 +32,66 @@
 #include "core/io/json.h"
 #include "core/os/copymem.h"
 #include "core/project_settings.h"
+#include "editor/editor_log.h"
 #include "editor/editor_node.h"
 
 GDScriptLanguageProtocol *GDScriptLanguageProtocol::singleton = NULL;
 
-void GDScriptLanguageProtocol::on_data_received(int p_id) {
-	lastest_client_id = p_id;
-	Ref<WebSocketPeer> peer = server->get_peer(p_id);
-	PoolByteArray data;
-	if (OK == peer->get_packet_buffer(data)) {
-		String message;
-		message.parse_utf8((const char *)data.read().ptr(), data.size());
-		if (message.begins_with("Content-Length:")) return;
-		String output = process_message(message);
-		if (!output.empty()) {
-			CharString charstr = output.utf8();
-			peer->put_packet((const uint8_t *)charstr.ptr(), charstr.length());
+void GDScriptLanguageProtocol::on_data_received(int bytes) {
+	// Check if header is big enough
+	if (bytes > 20) {
+		//Try to read "Content-Length:<num>\r\n\r\n" header
+		uint8_t *header_start = (uint8_t *)memalloc(16);
+		peer->get_data(header_start, 16);
+		bytes -= 16;
+		String header;
+		header.parse_utf8((const char *)header_start, 16);
+		memfree(header_start);
+		ERR_FAIL_COND(!header.begins_with("Content-Length: "));
+
+		// Read bytes until end of header
+		while (!header.ends_with("\r\n\r\n")) {
+			uint8_t *buff = (uint8_t *)memalloc(1);
+			peer->get_data(buff, 1);
+			bytes -= 1;
+			String str_buff;
+			str_buff.parse_utf8((const char *)buff, 1);
+			header += str_buff;
+			memfree(buff);
 		}
+
+		int content_length = header.substr(16, header.size()).to_int();
+		uint8_t *content = (uint8_t *)memalloc(content_length);
+		if (OK == peer->get_data(content, content_length)) {
+			String message;
+			message.parse_utf8((const char *)content, content_length);
+			String output = process_message(message);
+			if (!output.empty()) {
+				CharString charstr = output.utf8();
+				peer->put_data((const uint8_t *)charstr.ptr(), charstr.length());
+			}
+		}
+		memfree(content);
+	} else {
+		EditorNode::get_log()->add_message("Unable to parse header", EditorLog::MSG_TYPE_EDITOR);
+		uint8_t *data = (uint8_t *)memalloc(bytes);
+		peer->get_data(data, bytes); // Flush buffer
+		memfree(data);
 	}
 }
 
-void GDScriptLanguageProtocol::on_client_connected(int p_id, const String &p_protocal) {
-	clients.set(p_id, server->get_peer(p_id));
+void GDScriptLanguageProtocol::on_client_connected() {
```

Comment:
We should store a collection of peers instead of using just one. Then, when one connects, add it to the collection. When it disconnects, remove it from the collection